enlighten/scope/LaserControlFeature.py

```diff
@@ -521,8 +538,26 @@ def toggle_callback(self):
         # invert the previous state
         flag = not spec.settings.state.laser_enabled
         log.debug(f"toggle_callback: laser_enabled was {spec.settings.state.laser_enabled}, so setting {flag}")
-        self.set_laser_enable(flag)
 
+        # safety check for developmental firmware: if the user was trying to turn
+        # the laser off, turn it off regardless of previous "believed" state. The
+        # problem is there is no Checked() method on a button to determine 
+        # "previous state" (they're stateless), and it's possible that 
+        # state.laser_enabled could be wrong if wasatch.get_laser_enabled() isn't
+        # working right due to FW bugs, so go by our internal tracking of the
+        # user's "last request".
+        if flag:
+            # by default, we would be turning the laser ON at this point, since
+            # laser_enabled is apparently False
+
+            if spec.app_state.laser_state != LaserStates.DISABLED:
+                # however, laser_state implies that ENLIGHTEN thought the laser
+                # was either firing, or had been requested to fire
+
+                flag = False
+                log.critical(f"toggle_callback: turning laser OFF even though previous spec.settings.state.laser_enabled was {spec.settings.state.laser_enabled}, as spec.app_state.laser_state is {spec.app_state.laser_state}")
+
+        self.set_laser_enable(flag)
```

Comment:
This is still somewhat "live" developmental code, but has resolved issues I've seen with the current firmware. The long-term solution is to make the firmware robust and non-buggy, but for now we need to use ENLIGHTEN with "developmental" firmware and this reduces safety issues on the floor and in test labs.

Reply:
Actually, I think this addressed an intermittent laser issue CC observed in 220350 testing in the "released" ENLIGHTEN — not just for R&D.